### PR TITLE
fix: clean document service relation transforms

### PR DESCRIPTION
**Pull request by Marc-Roig**
### What does it do?

- Cleaned relation transformation methods on the document service (mainly to map documentIds )
- Provides an utility to map through relational data.

There are many ways to connect a relation:
```ts
id
[id]
[{id}]
{ set: id }
{ set: [id] }
{ set: [{id}] } // Object fields can contain other values, this is just an example
{ set: [{id}, id] } // yes, this is valid too  😓 
{ connect: id }
{ connect: [id] }
{ connect: [{id}] }
{ connect: [{id}, id] }
{ disconnect: id }
{ disconnect: [id] }
{ disconnect: [{id}] }
{ disconnect: [{id}, id] }
```

Mapping through all of the different permutations is clunky and difficult to maintain, the `mapRelation` utility always work with the same format:

```ts
mapRelation((relation) => {
	 // Mapps through all references of a relation, and always uses this same format
	const { id, documentId, locale, status } = relation
}, inputRelation)
```

Also, the `mapRelations` utility, standarises it to always return: 
```ts
{ set: [{id}] } // Object fields can contain other values, this is just an example
{ connect : [{id}] }
{ disconnect: [{id}] }
```

### CODE REVIEW

File: packages/core/core/src/services/document-service/transform/data.ts [hunk 1]
```diff
@@ -8,6 +8,8 @@ import { transformDataIdsVisitor as transformRelationDataIds } from './relations
 export const transformData = async (data: any, opts: any) => {
   const idMap = createIdMap({ strapi });
 
+  // Assign default locale to relations
+
```
Comment (innerdvations) [inline packages/core/core/src/services/document-service/transform/data.ts:12]: Is there supposed to be some code here?

**Review by innerdvations — COMMENTED** [review]
I haven't given it a thorough review, but looks good overall!

Looks like some of the tests you added are failing.

File: packages/core/core/src/services/document-service/transform/data.ts (hunk 1 above)
Comment (Marc-Roig) [reply packages/core/core/src/services/document-service/transform/data.ts:12]: oops, nope!

**Comment by Marc-Roig** [comment]
fixed the tests!

**Review by alexandrebodin — COMMENTED** [review]
I think you have too many commits in here you might want to rebase to only keep the mapRelation work ?

File: packages/core/core/src/services/document-service/transform/relations/extract/data-ids.ts [hunk 2]
```diff
@@ -1,126 +1,79 @@
...
+        const position = relation.position;
 
-        // If not connecting to any version on disabled d&p, we should connect to both draft and published relations at the same time
-        extractedIds.forEach((relation) => {
-          const targetLocale = getRelationTargetLocale(relation, {
-            targetUid: target as UID.Schema,
-            sourceUid: opts.uid,
-            sourceLocale: opts.locale,
-          });
+        if (position?.before) {
```
[the remark this answers is not in the document]
Comment (Marc-Roig) [reply packages/core/core/src/services/document-service/transform/relations/extract/data-ids.ts:65]: I might add a comment, but if user connects a relation like:

```
{ id: 'doc-id-1', position: { before: 'doc-id-2' } }
```

We need to map the positional argument to be an entry id too.

File: packages/core/core/src/services/document-service/transform/relations/transform/data-ids.ts [hunk 3]
```diff
@@ -1,235 +1,111 @@
...
+  return traverseEntityRelations(
+    async ({ key, value, attribute }, { set }) => {
+      // Find relational attributes, and return the document ids
+      const targetUid = attribute.target!;
+      const getIds = getRelationIds(idMap, targetUid, source);
 
-    const mapPosition = (relation: any) => {
-      if (isShortHand(relation) || !('position' in relation)) return relation;
+      // Transform the relation documentId to entity id
+      const newRelation = await mapRelation((relation) => {
```
[the remark this answers is not in the document]
Comment (Marc-Roig) [reply packages/core/core/src/services/document-service/transform/relations/transform/data-ids.ts:75]: ok! lets look at it together when we have a chance :)

File: packages/core/core/src/services/document-service/transform/relations/transform/default-locale.ts [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+    return data;
+  }
+
+  // Store the default locale to avoid multiple calls
+  let defaultLocale: string;
+
+  /**
+   * Traverse the entity input data and set the default locale to relations
+   */
+  return traverseEntityRelations(
```
[the remark this answers is not in the document]
Comment (Marc-Roig) [reply packages/core/core/src/services/document-service/transform/relations/transform/default-locale.ts:23]: good idea :) will do that

**Review by alexandrebodin — APPROVED** [review]
except the minor change the logic looks good and we can do the extra-transform stuff later :)

**Pull request merged**